### PR TITLE
Chapter 7: Sign up

### PR DESCRIPTION
**Pull request by phucynwa**

### CODE REVIEW

File: app/assets/stylesheets/custom.scss [hunk 1]
```diff
@@ -84,3 +84,81 @@ footer {
     }
   }
 }
+
+
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Please remove this extra line

File: app/assets/stylesheets/custom.scss [hunk 2]
```diff
@@ -84,3 +84,81 @@ footer {
...
+$gray-medium-light: #eaeaea;
+
+@mixin box_sizing {
+  -moz-box-sizing:    border-box;
+  -webkit-box-sizing: border-box;
+  box-sizing:         border-box;
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Only 1 space between key and valye

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -1,4 +1,25 @@
...
+  def show
+    @user = User.find(params[:id])
+  end
+
   def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: Please remove optional parenthesis ()

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -1,4 +1,25 @@
...
+    if @user.save
+      flash[:success] = "Welcome to the Sample App!"
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: Please use I18n(internationalization)
[http://guides.rubyonrails.org/i18n.html](http://guides.rubyonrails.org/i18n.html)

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -1,4 +1,25 @@
...
+      redirect_to @user
+    else
+      render "new"
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: Please use symbol :new instead of string "new"

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -1,4 +1,25 @@
...
+    end
   end
+
+  private
+
+    def user_params
+      params.require(:user).permit(:name, :email, :password, :password_confirmation)
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: Remove optional parenthesis. 
``
params.require(:user).permit :name, :email, :password, :password_confirmation
``

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -1,4 +1,25 @@
...
+      flash[:success] = "Welcome to the Sample App!"
+      redirect_to @user
+    else
+      render "new"
+    end
   end
+
+  private
+
+    def user_params
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: Here, no need 2 space indentation.

File: app/helpers/users_helper.rb [hunk 8]
```diff
@@ -1,2 +1,7 @@
 module UsersHelper
+  def gravatar_for(user, size: 80)
+    gravatar_id = Digest::MD5::hexdigest(user.email.downcase)
```
Comment (mdmahmudurrahman) [inline app/helpers/users_helper.rb]: Please remove optional ()

File: app/helpers/users_helper.rb [hunk 9]
```diff
@@ -1,2 +1,7 @@
 module UsersHelper
+  def gravatar_for(user, size: 80)
```
Comment (mdmahmudurrahman) [inline app/helpers/users_helper.rb]: Please remove optional ()

File: app/helpers/users_helper.rb [hunk 10]
```diff
@@ -1,2 +1,7 @@
...
+    gravatar_id = Digest::MD5::hexdigest(user.email.downcase)
+    gravatar_url = "https://secure.gravatar.com/avatar/#{gravatar_id}?s=#{size}"
+    image_tag(gravatar_url, alt: user.name, class: "gravatar")
```
Comment (mdmahmudurrahman) [inline app/helpers/users_helper.rb]: remove ()

File: app/views/layouts/application.html.erb [hunk 11]
```diff
@@ -1,4 +1,4 @@
-<!DOCTYPE html>
+us<!DOCTYPE html>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/application.html.erb]: What is us?

File: app/views/layouts/application.html.erb [hunk 12]
```diff
@@ -1,4 +1,4 @@
...
 <html>
 <head>
   <title><%= full_title(yield(:title)) %></title>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/application.html.erb]: remove optional ()
``
full_title yield(:title)
``

File: app/views/shared/_error_messages.html.erb [hunk 13]
```diff
@@ -0,0 +1,12 @@
+<% if @user.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      The form contains <%= pluralize(@user.errors.count, "error") %>.
```
Comment (mdmahmudurrahman) [inline app/views/shared/_error_messages.html.erb]: Remove optional ()

File: app/views/shared/_error_messages.html.erb [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+    </div>
+    <ul>
+    <% @user.errors.full_messages.each do |msg| %>
```
Comment (mdmahmudurrahman) [inline app/views/shared/_error_messages.html.erb]: 2 space indentation

File: app/views/static_pages/home.html.erb [hunk 15]
```diff
@@ -11,5 +11,4 @@
 
 </div>
 
-<%= link_to image_tag("rails.png", alt: "Rails logo"),
-'http://rubyonrails.org/' %>
+<%= link_to image_tag("rails.png", alt: "Rails logo"), 'http://rubyonrails.org/' %>
```
Comment (mdmahmudurrahman) [inline app/views/static_pages/home.html.erb]: Use double quote

File: app/views/users/new.html.erb [hunk 16]
```diff
@@ -1,2 +1,19 @@
-<h1><%= t "user" %></h1>
-<p>Find me in app/views/users/new.html.erb</p>
+<% provide(:title, "Sign up") %>
```
Comment (mdmahmudurrahman) [inline app/views/users/new.html.erb]: Remove ()

File: app/views/users/new.html.erb [hunk 17]
```diff
@@ -1,2 +1,19 @@
...
+
+<h1>Sign up</h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user, url: signup_path) do |f| %>
```
Comment (mdmahmudurrahman) [inline app/views/users/new.html.erb]: Remove ()

File: app/views/users/new.html.erb [hunk 18]
```diff
@@ -1,2 +1,19 @@
...
+      <%= render "shared/error_messages" %>
+      <%= f.label :name %>
+      <%= f.text_field :name, class: "form-control" %>
+      <%= f.label :email %>
+      <%= f.email_field :email, class: "form-control" %>
+      <%= f.label :password %>
+      <%= f.password_field :password, class: "form-control" %>
+      <%= f.label :password_confirmation, "Confirmation" %>
```
Comment (mdmahmudurrahman) [inline app/views/users/new.html.erb]: I18n

File: app/views/users/new.html.erb [hunk 19]
```diff
@@ -1,2 +1,19 @@
...
+      <%= f.password_field :password_confirmation, class: "form-control" %>
+      <%= f.submit "Create my account", class: "btn btn-primary" %>
```
Comment (mdmahmudurrahman) [inline app/views/users/new.html.erb]: I18n

File: app/views/users/show.html.erb [hunk 20]
```diff
@@ -0,0 +1,11 @@
+<% provide(:title, @user.name) %>
```
Comment (mdmahmudurrahman) [inline app/views/users/show.html.erb]: Remove ()

File: app/controllers/users_controller.rb [hunk 21]
```diff
@@ -1,4 +1,25 @@
...
+  end
+
   def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new user_params
+    if @user.save
+      flash[:success] = t "welcome"
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: Em, can you use I18n here
[http://guides.rubyonrails.org/i18n.html](http://guides.rubyonrails.org/i18n.html)

File: app/views/users/new.html.erb [hunk 22]
```diff
@@ -1,2 +1,19 @@
-<h1><%= t "user" %></h1>
-<p>Find me in app/views/users/new.html.erb</p>
+<% provide :title, I18n.t("signup") %>
```
Comment (mdmahmudurrahman) [inline app/views/users/new.html.erb]: In this section, you can study about more how you can use I18n in views
[http://guides.rubyonrails.org/i18n.html#abstracting-localized-code](http://guides.rubyonrails.org/i18n.html#abstracting-localized-code)

File: app/views/users/new.html.erb [hunk 23]
```diff
@@ -1,2 +1,19 @@
...
+
+<h1><%= I18n.t("signup") %></h1>
```
Comment (mdmahmudurrahman) [inline app/views/users/new.html.erb]: Same here

File: app/views/users/new.html.erb [hunk 24]
```diff
@@ -1,2 +1,19 @@
...
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @user, url: signup_path do |f| %>
+      <%= render "shared/error_messages" %>
+      <%= f.label :name %>
+      <%= f.text_field :name, class: "form-control" %>
+      <%= f.label :email %>
+      <%= f.email_field :email, class: "form-control" %>
+      <%= f.label :password %>
+      <%= f.password_field :password, class: "form-control" %>
+      <%= f.label :password_confirmation, I18n.t("confirmation") %>
```
Comment (mdmahmudurrahman) [inline app/views/users/new.html.erb]: In this section, you can study about more how you can use I18n in views
[http://guides.rubyonrails.org/i18n.html#abstracting-localized-code](http://guides.rubyonrails.org/i18n.html#abstracting-localized-code)

File: app/views/users/new.html.erb [hunk 25]
```diff
@@ -1,2 +1,19 @@
...
+      <%= f.password_field :password_confirmation, class: "form-control" %>
+      <%= f.submit I18n.t("create_my_account"), class: "btn btn-primary" %>
```
Comment (mdmahmudurrahman) [inline app/views/users/new.html.erb]: Same here

File: app/assets/stylesheets/custom.scss [hunk 26]
```diff
@@ -84,3 +84,80 @@ footer {
     }
   }
 }
+
+$gray-medium-light: #eaeaea;
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Remove this color variable to an external file, then import it from there.

File: app/assets/stylesheets/custom.scss [hunk 27]
```diff
@@ -84,3 +84,80 @@ footer {
...
+  float: left;
+  margin-right: 10px;
+}
+
+.gravatar_edit {
+  margin-top: 15px;
+}
+
+input, textarea, select, .uneditable-input {
+  border: 1px solid #bbb;
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Use color variables.

File: app/helpers/users_helper.rb [hunk 28]
```diff
@@ -1,2 +1,7 @@
 module UsersHelper
+  def gravatar_for user, size: 80
```
Comment (mdmahmudurrahman) [inline app/helpers/users_helper.rb]: Please use the config gem to manage the constants. 
For example, like the way you used it in the User model.

File: app/views/shared/_error_messages.html.erb [hunk 29]
```diff
@@ -0,0 +1,12 @@
+<% if @user.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      The form contains <%= pluralize @user.errors.count, "error" %>.
```
Comment (mdmahmudurrahman) [inline app/views/shared/_error_messages.html.erb]: Use I18n here.

File: app/models/user.rb [hunk 30]
```diff
@@ -1,7 +1,10 @@
 class User < ApplicationRecord
+  before_save { email.downcase! }
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Remove white space after the start and before the end of curly brace {}

File: app/views/static_pages/home.html.erb [hunk 31]
```diff
@@ -1,5 +1,5 @@
 <div class="center jumbotron">
-	<h1>Welcome to the Sample App</h1>
+	<h1><%= t :welcome %></h1>
 
 	<h2>
 		This is the home page for the
```
[the remark this answers is not in the document]
Comment (phucynwa) [reply app/views/static_pages/home.html.erb]: Isn't it I18n ?

File: app/controllers/users_controller.rb [hunk 32]
```diff
@@ -1,4 +1,25 @@
 class UsersController < ApplicationController
+  def show
+    @user = User.find params[:id]
```
[the remark this answers is not in the document]
Comment (phucynwa) [reply app/controllers/users_controller.rb]: I knew it. But what should I do ?

**Pull request merged**